File: go-version/leetcode/LC2597/beautiful_subsets.go
package LC2597

/**
 * @Author: cutejiuge cutejiuge@163.com
 * @Date: 2025/3/7 上午9:18
 * @FilePath: beautiful_subsets
 * @Description: 美丽子集，暴力回溯
 */

func beautifulSubsets(nums []int, k int) int {
	mp := map[int]int{}
	ans := -1
	var backtrack func(int, []int, int, map[int]int)
	backtrack = func(idx int, nums []int, k int, m map[int]int) {
		//枚举终结
		if idx == len(nums) {
			ans++
			return
		}
		//不选择当前直接跳过
		backtrack(idx+1, nums, k, m)
		//判断能选就选中
		cur := nums[idx]
		if m[cur-k] == 0 && m[cur+k] == 0 { //可选
			m[cur]++
			backtrack(idx+1, nums, k, m)
			//恢复现场
			m[cur]--
		}
	}
	backtrack(0, nums, k, mp)
	return ans
}

File: python-version/LC2597/beautiful_subsets.py
class Solution:
    def __init__(self):
        self.__nums = None
        self.__k = None
        self.__map = None
        self.__ans = -1

    def beautifulSubsets(self, nums: list[int], k: int) -> int:
        self.__nums = nums
        self.__k = k
        self.__map = {}
        self.__ans = -1
        self.backtrack(0)
        return self.__ans

    def backtrack(self, idx: int):
        if idx == len(self.__nums):
            self.__ans += 1
            return
        # 不选择当前元素，直接往下走
        self.backtrack(idx+1)
        # 如果能选就选中
        cur = self.__nums[idx]
        if self.__map.get(cur - self.__k, 0) == 0 and self.__map.get(cur + self.__k, 0) == 0:
            self.__map[cur] = self.__map.get(cur, 0) + 1
            self.backtrack(idx+1)
            # 恢复现场
            self.__map[cur] = self.__map.get(cur, 0) - 1
